pkg/pillar/cmd/domainmgr/containerd.go

```diff
@@ -90,70 +94,159 @@ func ctrRun(domainName, xenCfgFilename, imageHash string, envList map[string]str
-	return nil
+func ctrCreate(containerID, imageName string, status *types.DomainStatus) (string, error) {
+	var (
+		ociOpts       []oci.SpecOpts
+		containerOpts []containerd.NewContainerOpts
+		ociSpec       oci.Spec
+		err           error
+	)
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return "", fmt.Errorf("Exception while loading container. %v", err)
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	if len(status.EnvVariables) > 0 {
+		envVars := buildEnvVarsSlice(status.EnvVariables)
+		ociOpts = append(ociOpts, oci.WithEnv(envVars))
+	}
+
+	if len(ociOpts) > 0 {
+		containerOpts = append(containerOpts, containerd.WithSpec(&ociSpec, ociOpts...))
+	}
+	image, err := getImage(imageName)
+	if err != nil {
+		return "", fmt.Errorf("Unable to fetch image %v. %v", imageName, err.Error())
+	}
+
+	containerSnapshot := fmt.Sprintf("%s-snapshot", containerID)
+	containerOpts = append(containerOpts, containerd.WithImage(image))
+	containerOpts = append(containerOpts, containerd.WithNewSnapshot(containerSnapshot, image))
+	containerOpts = append(containerOpts, containerd.WithNewSpec(oci.WithImageConfig(image)))
+	_, err = ctrdClient.NewContainer(
+		ctrdCtx,
+		containerID,
+		containerOpts...,
+	)
+	if err != nil {
+		log.Errorf("Could not build new containerd container: %v. %v", containerID, err.Error())
+		return "", fmt.Errorf("Could not build new containerd container: %v. %v", containerID, err.Error())
+	}
+
+	_, err = ctrCreateNewTask(containerID)
+	if err != nil {
+		log.Errorf("Could not create a new task for container: %v. %v", containerID, err.Error())
+		return "", fmt.Errorf("Could not create a new task for container: %v. %v", containerID, err.Error())
+	}
+	return containerID, nil
+}
+
+// ctrCreateNewTask - Creates a new task for the container.
+// This step is responsible for setting up rootfs of the container under /run/containerd/<RunTime>/<NameSpace>/<Container id>/
+// NOTE: As per containerd's flow, to start a container, first a task is created which sets up container's rootfs,
+// then the task is started which invokes runc binary with the rootfs. Since we are not using runc, we need to stop at task create step.
+func ctrCreateNewTask(containerID string) (containerd.Task, error) {
+	var err error
+	ctrdClient, err := getContainerdClient(ctrdSocket)
+	if err != nil {
+		log.Errorf("Could not build containerd client. %v", err.Error())
+		return nil, fmt.Errorf("Exception while loading container. %v", err)
+	}
+	defer ctrdClient.Close()
+	ctrdCtx := getContainerdContext(CtrdServicesNamespace)
+
+	container, err := loadContainer(containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	return container.NewTask(ctrdCtx, cio.NewCreator(cio.WithStdio))
 }
 
 // ctrPrepare prepare an existing container
 func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
```

Comment:
Same here. The only thing we use `status` for is to get:

* `status.DisplayName string`
* `status.EnvVariables map[string]string`

If we pass those 2 explicitly to `ctrPrepare()`, then the func has zero eve dependencies, and becomes much easier to test.

Reply:
Actually, take it one step further. Here is the beginning of `ctrPrepare()`:

```go
func ctrPrepare(containerImageSha256 string, status *types.DomainStatus) (string, error) {
	ociFilename, err := utils.VerifiedImageFileLocation(containerImageSha256)
 	log.Infof("ociFilename %s sha %s", ociFilename, containerImageSha256)
```

We only use `containerImageSha256` to calculate the `ociFilename`. Move that back out to the calling function in `domainmgr.go`, passing it just the filename, and we can get rid of importing `lf-edge/eve/pkg/pillar/utils`. Then there will be zero dependencies.

Reply:
Done.